src/beamlime/constructors/inspectors.py

```diff
@@ -37,26 +37,38 @@ def validate_annotation(annotation: Any) -> Literal[True]:
 
 
 Product = TypeVar("Product")
+
+
+def _is_type_newtype_py310(tp: Any) -> bool:
+    from typing import NewType
+
+    return isinstance(NewType, type) and isinstance(tp, NewType)
+
+
 _NewT = NewType("_NewT", int)
 
 
+def _is_type_newtype_py39(tp: Any) -> bool:
+    return (
+        callable(tp)
+        and hasattr(tp, "__supertype__")
+        and _NewT.__module__ == tp.__module__
+        and _NewT.__qualname__ == tp.__qualname__
+    )
+
+
 class NewTypeMeta(type):
     def __instancecheck__(cls, instance: Any) -> bool:
-        return (
-            callable(instance)
-            and hasattr(instance, "__supertype__")
-            and _NewT.__module__ == instance.__module__
-            and _NewT.__qualname__ == instance.__qualname__
-        )
+        return _is_type_newtype_py310(instance) or _is_type_newtype_py39(instance)
```

Comment:
Also, have a look at this: https://github.com/pydantic/pydantic/blob/c83e5fcd2da5693d4dbcf031f749be66a81b322d/pydantic/_internal/_typing_extra.py#L118

I'm not saying your solution is wrong, but theirs looks simpler.

Reply:
Yeah it's true, I needed to check the ``.__module__`` and ``.__qualname__`` because checking ``__supertype__`` was not sufficient in some cases though... I'll just keep it this way for now. It'll be unnecessary from py3.10 anyways...